### ISSUE TITLE
Separate `backward` and `step` in `deepspeed` wrapper

### ISSUE THREAD

**Issue by ummagumm-a** [huggingface/accelerate#3303]
Hello. I have a request to separate the two operations. Right now it prohibits from using a feature from deepspeed to access gradients in a model. See this issue: https://github.com/microsoft/DeepSpeed/issues/5199#issuecomment-2545973162.

My specific use case is fine-tuning a subset of weights in embedding and lm_head modules when initializing new tokens. I wanted to implement this logic in [on_pre_optimizer_step](https://huggingface.co/docs/transformers/en/main_classes/callback#transformers.TrainerCallback.on_pre_optimizer_step) callback in `transformers` library, which is right in between of `backward` and `step`.

There I want to use [safe_get_full_grad](https://deepspeed.readthedocs.io/en/latest/zero3.html#deepspeed.utils.safe_get_full_grad) from deepspeed to get the gradients. It has a condition that it should be called after `backward` and before `step`. But since `accelerator.backward` [(usage link in transformers.trainer training loop)](https://github.com/huggingface/transformers/blob/da334bcfa8ff7feb85138ce90ca7340e4fc6e704/src/transformers/trainer.py#L3702) has these two operations merged ([here](https://github.com/huggingface/accelerate/blob/200c9eb7833cfa505907f6f224ebf5a275aa6d92/src/accelerate/utils/deepspeed.py#L255)) when used with `deepspeed`, using `safe_get_full_grad` is impossible. 

Here is a [small example](https://gist.github.com/ummagumm-a/feef6e25771e4fd9dc3a90d8449e343f) to reproduce this issue. Please see the link attached earlier in deepspeed issues to see a discussion about it.

**Comment by ummagumm-a** [comment]
@tjruwase fyi
